keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,577 @@
+
+### Network Context
+
+The minimal NetworkContext looks like the following struct in golang:
+
+```go
+type NetworkContext struct {
+        // NetworkName is the unique name of the network 
+        // to select to route the traffic to. 
+        NetworkName string
+}
+```
+
+NetworkName specifies the network to route traffic to. 
+The KAS starts with a list of registered network names and 
+the corresponding Connectivity services. So the KAS knows where to 
+proxy the traffic to, otherwise it return an “Unknown network” error. 
+
+The KAS starts with a proxy configuration like the below example. 
+The example specifies 4 networks. "direct" specifies the KAS talking directly 
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy 
+listening at 1.2.3.5:5679. While these are represented as resources 
+they are not intended to be loaded dynamically. 
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.ca
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.ca
+
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing 
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+    // Proxy connects to a remote address by stream id, and establish
+    // a bi-directional stream of packet exchange.
+    rpc Proxy(stream Packet) returns (stream Packet) {}
+}
+
+enum PacketType {
+  DIAL_REQ = 0;
+  DIAL_RSP = 1;
+  CLOSE_REQ = 2;
+  CLOSE_RSP = 3;
+  DATA = 4;
+}
+
+message Packet {
+  PacketType type = 1;
+
+  oneof payload {
+    DialRequest dialRequest = 2;
+    DialResponse dialResponse = 3;
+    CloseRequest closeRequest = 4;
+    CloseResponse closeResponse = 5;
+    Data data = 6;
+  }
+}
+
+// Error is sent when error happens from remote side of connection when it tries
+// to read or write from it
+enum Error {
+  EOF = 0;
+  EIO = 1;
+  ECONNRESET = 2;
+  ETIMEOUT = 3;
+  EADDRNOTAVAIL = 4;
+  EMFILE = 5;
+  // … 
+}
+
+// DialRequest represents a request to dial to an address on the other 
+// side of the tunnel. The format is inspired by golang's net interface
+// https://golang.org/pkg/net/#Dial 
+message DialRequest {
+    // network representing a named network. "Tcp" is the only supported
```

Comment:
Do you plan to support more than just TCP?

Reply:
Right, maybe UDP, not the highest priority at this point of time tho.

Reply:
The API server doesn't do anything with UDP today. For a while now, we've discussed breaking the proxy functionality out of the API server. I would be pretty hesitant to add more functionality at this point.

Reply:
exactly, to be clear, we don't have a plan to support UDP. the field is there just for the sake of extensibility. it's okay to me to remove it and assume all traffic is tcp, but in future if there is need, there will be a proto change.

Reply:
Adding a field later that defaults to tcp is a backwards compatible change.